File: dynamicProgramming/2.Tabulation/4.howSum.py
def howSum(targetSum, numbers):
	table = []
	isPossible = []
	for i in range(targetSum + 1):
		table.append([])
		isPossible.append(False)
	# print(table)
	

	# If targetsum is 0, it is always possible to
	# we can just return empty list []

	isPossible[0] = True
	# print(isPossible)
	for i in range(targetSum + 1):
		for j in range(len(numbers)):
			if isPossible[i]:
				# What is the value that we will be getting?
				value = table[i]
				print("Value before: ")
				print(value)
				value.append(numbers[j])
				print("Value after append: ")
				print(value)
				sumValue = sum(value)
				print("Sum :" + str(sumValue))
				if sumValue <= targetSum:
					isPossible[sumValue] = True
					table[sumValue] = value
					print("Table value after assigning: ")
					print(table)
	if not isPossible[targetSum]:
		return None
	else:
		return table[targetSum]


print(howSum(1, [1])) # True
# print(howSum(2,[1])) # True
# print(howSum(3,[2])) # False
# print(howSum(4,[1,2])) # True
# print(howSum(7, [2,3])) # True
# print(howSum(7, [5, 3, 4, 7])) # True
# print(howSum(7, [2, 4])) # False
# print(howSum(8, [2,3, 5])) # True
# print(howSum(300, [7,14])) # False

File: dynamicProgramming/2.Tabulation/4.howSum.cpp
#include<bits/stdc++.h>
using namespace std;

void printTable(vector< vector<int> > table) {
	cout << "Table value is: " << endl;
	for(int i=0; i<table.size(); i++) {
		cout << "Table index: " << i << " | " << table[i].size() << endl;
		for(int j=0; j<table[i].size(); j++){
			cout << table[i][j] << " ";
		}
		cout << endl;
	}
	cout << endl;
}

int getSum(vector<int> arr) {
	int sum = 0;
	for(int i=0; i<arr.size(); i++){
		sum = sum + arr[i];
	}

	return sum;
}

vector<int> howSum(int targetSum, vector<int> numbers) {
	// We create a table
	vector< vector<int> > table(targetSum + 1, vector<int>());
	vector<bool> isPossible(targetSum + 1, false);

	// targetSum of 0 is always possible
	isPossible[0] = true;

	for(int i=0; i<= targetSum; i++){
		if(isPossible[i]){
			for(int j=0; j<numbers.size(); j++){
				vector<int> value = table[i];
				value.push_back(numbers[j]);
				int sumValue = getSum(value);

				if(sumValue <= targetSum){
					isPossible[sumValue] = true;
					table[sumValue] = value;
				}
			}
		}
	}

	return table[targetSum];
}

int main(){
	int t;
	cin >> t;
	cout << "Number of test Cases is: " << t << endl;
	for(int z=0; z<t; z++){
		int targetSum;
		int numbersLength;
		cin >> targetSum >> numbersLength;
		cout << "Target Sum is: " << targetSum << endl;
		cout << "Length of numbers array is: " << numbersLength << endl;
		vector<int> numbers(numbersLength);
		cout << "Numbers list is: " << endl;
		for(int i=0; i<numbersLength; i++){
			int temp;
			cin >> temp;
			numbers[i] = temp;
			cout << numbers[i] << " ";
		}
		cout << endl;
		vector<int> result;
		result = howSum(targetSum, numbers);
		cout << "Result: " << endl;
		for(int i=0; i<result.size(); i++){
			cout << result[i] << " ";
		}
		cout << endl;
		cout << endl;
	}
}